STORM-845 Storm ElasticSearch connector

sweetest opened the pull request:
This PR is about introducing another external module, storm-elasticsearch.

At this moment, it contains two basic bolt, namely, EsIndexBolt and EsPercolatorBolt, and trident state, also basic test cases for them.

It will be very nice to have it included in Storm, since ElasticSearch is one of the greatest open source search & storage engine which is widely used.

Please review this and merge to master.

harshach commented:
@sweetest it looks good on my first look. I volunteer to be a sponsor for this connector. Can you upmerge this. I'll work on getting this in. Thanks.

harshach commented on external/storm-elasticsearch/README.md:
```diff
@@ -0,0 +1,71 @@
+|Arg  |Description | Type
+|---	|--- |---
+|clusterName | ElasticSearch cluster name | String (required) |
+|host | ElasticSearch host | String array (required) |
```
having a hostname and port as separate arrays confusing . How about hostname:port as the string itself and users can host1:port1, host2:port2

HeartSaVioR replied:
+1 on @harshach opinion.

sweetest replied:
@harshach Thanks for the review, I'll work on it.

sweetest commented:
@harshach I've addressed your comment and rebased based on current upstream master.
Please have a look and let me know if there's any other issues.

mnogu commented on external/storm-elasticsearch/README.md:
```diff
@@ -0,0 +1,66 @@
+# Storm ElasticSearch Bolt & Trident State
```
This search engine was ElasticSearch as you wrote, but is Elasticsearch now.
In short, the 2nd "s" is a small "s".
https://www.elastic.co/products/elasticsearch

sweetest replied:
@mnogu Thanks for your opinion, I'll fix it too.

harshach commented:
@sweetest thanks for the quick turnaround. I'll take a look.

mnogu commented:
Looks good.

HeartSaVioR commented:
Sorry it feels you off the topic, but recently I found elasticsearch-hadoop supports Storm officially which is maintained by Elastic.
https://github.com/elastic/elasticsearch-hadoop

I'd like to see a benefit to have this to Storm's external module instead of using Elastic's official module.
@sweetest Could you take a look at elasticsearch-hadoop?

If we're sure about the benefit, I also volunteer to be a sponsor, and take a look deeply.

mnogu commented:
Elasticsearch-hadoop doesn't support Trident API.

HeartSaVioR commented:
@mnogu 
Yeah, right. My first look this PR seems superior than elasticsearch-hadoop.

Actually I'm thinking about future maintenance, where is the best place to improve Elasticsearch connector, Storm Project itself, or Elastic. 
If we think elasticsearch-hadoop is more proper place, we can contribute elasticsearch-hadoop to improve. If we don't think, we can adopt this module and maintain it.

@sweetest If you don't mind, I'd like to hear your opinion. Thanks for contributing!

revans2 commented:
@sweetest The code looks great to me.  I would like to see some more javadocs explaining what each of the bolts/state do/need as input.  I would also like to see some javadocs around beginCommit and commit in the state explaining that adding a document to the search index is an idempotent operation so the commits can be ignored.

sweetest commented:
Thanks for comments from all of you guys.

@harshach 
I added javadocs you've mentioned.

@HeartSaVioR @mnogu 
Regarding Elasticsearch-hadoop, those bolt/spout interfaces are for elasticsearch-hadoop, not for native elasticsearch. Thus target users may differ and this could be one reason for maintaining separate connector here. But more importantly, I think it's beneficial to have ES driver in storm repo, since having various 3rd party integrations would help attract possible users into Storm community.

HeartSaVioR commented:
@sweetest 
Good. 
I also think supporting ES is a valuable feature since many users already enjoys with ES + Kibana.
And at a glimpse EsSpout and EsBolt contains only very basic feature so we may be eager for having more powerful ES + Storm connector.

I'd also like to volunteer to be a sponsor this module. I'll take a look and comment later.

HeartSaVioR commented on external/storm-elasticsearch/README.md:
```diff
@@ -0,0 +1,66 @@
+EsIndexBolt streams tuples directly into Elasticsearch. Tuples are indexed in specified index & type combination. 
+User should make sure that there are "source", "index","type", and "id" fields declared in preceding bolts or spout.
+"index" and "type" fields are used for identifying target index and type.
+"source" is a document in JSON format string that will be indexed in elastic search.
```
elastic search => Elasticsearch

HeartSaVioR commented on external/storm-elasticsearch/README.md:
```diff
@@ -0,0 +1,66 @@
+EsConfig esConfig = new EsConfig();
+esConfig.setClusterName(clusterName);
+esConfig.setNodes(new String[]{"localhost:9300"});
+EsIndexBolt indexBolt = new IndexBolt(esConfig);
```
new IndexBolt(esConfig) => new EsIndexBolt(esConfig)

HeartSaVioR commented on external/storm-elasticsearch/README.md:
````diff
@@ -0,0 +1,66 @@
+EsPercolateBolt percolateBolt = new EsPercolateBolt(esConfig);
+```
+
+### EsConfig (org.apache.storm.elasticsearch.common.EsConfig)
````
### EsConfig => ## EsConfig
(EsConfig is not only for EsPercolateBolt.)

sweetest replied:
addressed as you've commented.

HeartSaVioR commented on external/storm-elasticsearch/README.md:
````diff
@@ -0,0 +1,66 @@
+   esConfig.setNodes(new String[]{"localhost:9300"});
+  ```
+
+EsConfig params
````
```EsConfig params => ### EsConfig params```

It would be better to emphasize this section.

sweetest replied:
addressed as you've commented.

HeartSaVioR commented on external/storm-elasticsearch/src/main/java/org/apache/storm/elasticsearch/bolt/EsIndexBolt.java:
```diff
@@ -0,0 +1,71 @@
+            client.prepareIndex(index, type, id).setSource(source).execute().actionGet();
+            collector.ack(tuple);
+        } catch (Exception e) {
+            e.printStackTrace();
```
Please use LOG instead of printing to stderr.

sweetest replied:
removed this part.

HeartSaVioR commented on external/storm-elasticsearch/src/main/java/org/apache/storm/elasticsearch/bolt/EsPercolateBolt.java:
```diff
@@ -0,0 +1,81 @@
+            }
+            collector.ack(tuple);
+        } catch (Exception e) {
+            e.printStackTrace();
```
Please use LOG instead of printing to stderr.
(Line 71)

sweetest replied:
removed this part.

HeartSaVioR commented on external/storm-elasticsearch/src/main/java/org/apache/storm/elasticsearch/trident/EsStateFactory.java:
```diff
@@ -0,0 +1,51 @@
+import java.util.Map;
+
+public class EsStateFactory implements StateFactory {
+    private static final Logger LOG = LoggerFactory.getLogger(EsStateFactory.class);
```
LOG => redundency, can be removed

sweetest replied:
addressed as you've commented.

HeartSaVioR commented on external/storm-elasticsearch/src/main/java/org/apache/storm/elasticsearch/bolt/EsPercolateBolt.java:
```diff
@@ -0,0 +1,81 @@
+            if (response.getCount() > 0) {
+                for (PercolateResponse.Match match : response) {
+                    String id = match.getId().toString();
+                    collector.emit(new Values(id));
```
I'm curious that emitting only query id list covers general use case. 
If users need to know query's information, they should write codes to retrieve it.
If PercolateResponse.Match can be serialized, I think it's better to emit it.

sweetest replied:
I modified code to emit source document(String) and Match.
So that one can search for original query document afterwards using information in Match object.

HeartSaVioR commented:
@sweetest 
The code looks great to me overall, too.

Btw, actually I read your post from your company's official technical blog.
(http://d2.naver.com/helloworld/1044388)
This module covers your use case, but IMO some functionalities should be added to cover more (and general) use cases.

Surely, we can just add to another issues, and implement it later. Here's the list.

- Introduces Tuple -> ES document mapper to get rid of constant field mapping (source, index, type)
- Introduces Lookup(or Query)Bolt which retrieves matched documents from ES and emit them
- Introduces BaseQueryFunction which is same to Lookup(or Query)Bolt but can be used in Trident topology

For Lookup(or Query)Bolt and BaseQueryFunction, we need to introduce ES document -> Tuple mapper, and we may want to set which field is used for querying. 
We may also want to consider about the strategy - what we can do when too many documents are matched.

@revans2 @harshach @sweetest 
Please comment me when you're unsure about features I've suggested. Thanks!

HeartSaVioR commented:
@sweetest 
And I saw elasticsearch-hadoop has EsSpout, but this module doesn't include Spout implementation.
https://github.com/elastic/elasticsearch-hadoop/blob/master/storm/src/main/java/org/elasticsearch/storm/EsSpout.java

Did you take a look at EsSpout and decided it is not necessary for current module?

sweetest commented:
I addressed comments @HeartSaVioR made.
As you've said, I started from where I'm more experienced so this module does not contain comprehensive functionality. I think we can discuss and expand more afterwards, definitely.
(Thanks for having read the article BTW :) )

For spout implementation, I also saw one in elastic-hadoop. But I thought at that time that it would not be a meaningful feature since Elasticsearch is more of a storage and search engine, not a queue and thus it's not really optimized for implementing spout of it.

HeartSaVioR commented:
@sweetest 
Please check comments which doesn't show as outdated. You seems missed some comments.
Other things are fine. 

And I agree that ES is not a queue so it may not feel comfortable to implement Spout for ES.

Overall I'm fine for merging this in. 
As @harshach want to volunteer this module, I'd like @harshach to take a look, too.
When @harshach confirms it's good to merge, I'll merge to master.

And please add me and @harshach as Committer Sponsors.
You can refer https://github.com/apache/storm/blob/master/external/storm-redis/README.md to see how to add.

Thanks for contributing!

sweetest commented:
All comments addressed, please check.

Thanks for review and all the help @HeartSaVioR , @harshach

HeartSaVioR commented on external/storm-elasticsearch/README.md:
```diff
@@ -0,0 +1,72 @@
+  
+## Committer Sponsors
+
+ * Robert Evans ([@revans2](https://github.com/revans2))
```
@sweetest 
Committer who volunteered to be sponsor is @harshach (Sriharsha Chintalapani), not @revans2 (Robert Evans).
@revans2 Do you want to volunteer to this module, too?

sweetest replied:
oh sorry I'll edit it right away

HeartSaVioR commented:
+1.
@harshach Could you take a look? Or do you want me to merge it without your review?

HeartSaVioR commented:
Maybe @harshach doesn't have time to review at the moment.
I'll merge it right now.

HeartSaVioR commented:
@sweetest 
I filed issues which I stated to JIRA, here's a list.

- https://issues.apache.org/jira/browse/STORM-974
- https://issues.apache.org/jira/browse/STORM-978
- https://issues.apache.org/jira/browse/STORM-979

I hope that you'd like to work on improving storm-elasticsearch more amazing.
